### PR TITLE
When the consumer read timeout, try to read after the value of `timeout`

### PR DESCRIPTION
**Pull request by yunnysunny**
The pull request is aimed to resolve the issue of https://github.com/Blizzard/node-rdkafka/issues/709 .

### CODE REVIEW

**Comment by yunnysunny** [comment]
You can restart the travis-ci, it can recovery from error.

**Comment by giulliano-bueno** [comment]
This PR seems to be good enough to be merged.

**Comment by iradul** [comment]
I think default value for this timeout is `1000` so this can be a breaking change for some. It would be much better to have a separate timeout field just for handling this scenario.

**Comment by yunnysunny** [comment]
Add a field of `m_retry_read_ms` to set the delay time before next read operation.

File: lib/kafka-consumer.js [hunk 1]
```diff
@@ -17,7 +17,8 @@ var KafkaConsumerStream = require('./kafka-consumer-stream');
 var LibrdKafkaError = require('./error');
 var TopicPartition = require('./topic-partition');
 var shallowCopy = require('./util').shallowCopy;
-
+var DEFAULT_RETRY_READ_INTERVAL = 500;
```
Comment (iradul) [inline lib/kafka-consumer.js:20]: Let's call this `CONSUME_LOOP_TIMEOUT_DELAY`

File: lib/kafka-consumer.js [hunk 2]
```diff
@@ -124,15 +125,24 @@ function KafkaConsumer(conf, topicConf) {
...
 KafkaConsumer.prototype.setDefaultConsumeTimeout = function(timeoutMs) {
-  this._consumeTimeout = timeoutMs;
+  this._consumeTimeout = Number(timeoutMs) || DEFAULT_CONSUME_TIME_OUT;
+};
+
+/**
+ * Set the default value of waiting for the next time to read when current reading get none data.
+ * @param {number} intervalMs - number of milliseconds to sleep before the next reading
+ */
+KafkaConsumer.prototype.setDefaultRetryReadInterval = function(intervalMs) {
```
Comment (iradul) [inline lib/kafka-consumer.js:144]: Please rename method name to `setDefaultConsumeLoopTimeoutDelay`

File: lib/kafka-consumer.js [hunk 3]
```diff
@@ -124,15 +125,24 @@ function KafkaConsumer(conf, topicConf) {
...
  */
 KafkaConsumer.prototype.setDefaultConsumeTimeout = function(timeoutMs) {
-  this._consumeTimeout = timeoutMs;
+  this._consumeTimeout = Number(timeoutMs) || DEFAULT_CONSUME_TIME_OUT;
+};
+
+/**
+ * Set the default value of waiting for the next time to read when current reading get none data.
+ * @param {number} intervalMs - number of milliseconds to sleep before the next reading
+ */
```
Comment (iradul) [inline lib/kafka-consumer.js:143]: Maybe reword like this
```
/**
 * Set the default sleep delay for the next consume loop after the previous one has timed out.
 * @param {number} intervalMs - number of milliseconds to sleep after a message fetch has timed out
 */
```

File: lib/kafka-consumer.js [hunk 4]
```diff
@@ -124,15 +125,24 @@ function KafkaConsumer(conf, topicConf) {
...
+KafkaConsumer.prototype.setDefaultRetryReadInterval = function(intervalMs) {
+  this._retryReadInterval = Number(intervalMs) || DEFAULT_RETRY_READ_INTERVAL;
```
Comment (iradul) [inline lib/kafka-consumer.js:145]: Please rename `_retryReadInterval` to `_consumeLoopTimeoutDelay`

File: src/kafka-consumer.cc [hunk 5]
```diff
@@ -1066,14 +1066,23 @@ NAN_METHOD(KafkaConsumer::NodeConsumeLoop) {
   } else {
     timeout_ms = static_cast<int>(maybeTimeout.FromJust());
   }
```
Comment (iradul) [inline src/kafka-consumer.cc:1068]: Above these lines, since there are now `3` arguments, validation needs to be fixed:
```
  if (!info[0]->IsNumber()) {
    return Nan::ThrowError("Need to specify a timeout");
  }

  if (!info[1]->IsNumber()) {
    return Nan::ThrowError("Need to specify a sleep delay");
  }

  if (!info[2]->IsFunction()) {
    return Nan::ThrowError("Need to specify a callback");
  }
```

File: src/workers.cc [hunk 6]
```diff
@@ -420,10 +420,12 @@ void KafkaConsumerDisconnect::HandleErrorCallback() {
 
 KafkaConsumerConsumeLoop::KafkaConsumerConsumeLoop(Nan::Callback *callback,
                                      KafkaConsumer* consumer,
-                                     const int & timeout_ms) :
+                                     const int & timeout_ms,
+                                     const int & retry_interval_ms) :
```
Comment (iradul) [inline src/workers.cc:424]: Please rename to `timeout_sleep_delay_ms`

File: src/workers.h [hunk 7]
```diff
@@ -288,6 +288,7 @@ class KafkaConsumerConsumeLoop : public MessageWorker {
   NodeKafka::KafkaConsumer * consumer;
   const int m_timeout_ms;
   unsigned int m_rand_seed;
+  const int m_retry_read_ms;
```
Comment (iradul) [inline src/workers.h:291]: Please rename to `m_timeout_sleep_delay_ms`

File: lib/kafka-consumer.js (hunk 4 above)
Comment (iradul) [inline lib/kafka-consumer.js:145]: Also I think you should not wrap interval within `Number(..)` because validation is done in c++ code

File: lib/kafka-consumer.js [hunk 8]
```diff
@@ -124,15 +125,24 @@ function KafkaConsumer(conf, topicConf) {
...
+  this._retryReadInterval = DEFAULT_RETRY_READ_INTERVAL;
 }
 
 /**
  * Set the default consume timeout provided to c++land
  * @param {number} timeoutMs - number of milliseconds to wait for a message to be fetched
  */
 KafkaConsumer.prototype.setDefaultConsumeTimeout = function(timeoutMs) {
-  this._consumeTimeout = timeoutMs;
+  this._consumeTimeout = Number(timeoutMs) || DEFAULT_CONSUME_TIME_OUT;
```
Comment (iradul) [inline lib/kafka-consumer.js:137]: Passing `0` here will not work, also you should not wrap interval within `Number(..)` because:
a) validation is done in c++ code
b) it's kind of breaking change because calling `setDefaultConsumeTimeout('anything but number')` will suddenly start working

Maybe something like `this._consumeTimeout = timeoutMs === 0 ? timeoutMs : timeoutMs || DEFAULT_CONSUME_TIME_OUT;` would be safer.

Comment (yunnysunny) [reply lib/kafka-consumer.js]: Set the value of this._consumeTimeout to timeoutMs directly.

File: lib/kafka-consumer.js [hunk 9]
```diff
@@ -124,7 +125,8 @@ function KafkaConsumer(conf, topicConf) {
   this.globalConfig = conf;
   this.topicConfig = topicConf;
 
-  this._consumeTimeout = 1000;
+  this._consumeTimeout = DEFAULT_CONSUME_LOOP_TIMEOUT_DELAY;
```
Comment (iradul) [inline lib/kafka-consumer.js:128]: Here you should use `DEFAULT_CONSUME_TIME_OUT` instead of `DEFAULT_CONSUME_LOOP_TIMEOUT_DELAY`

File: lib/kafka-consumer.js [hunk 10]
```diff
@@ -369,7 +379,7 @@ KafkaConsumer.prototype.unsubscribe = function() {
  * is fetched.
  */
 KafkaConsumer.prototype.consume = function(number, cb) {
-  var timeoutMs = this._consumeTimeout || 1000;
+  var timeoutMs = this._consumeTimeout || DEFAULT_CONSUME_LOOP_TIMEOUT_DELAY;
```
Comment (iradul) [inline lib/kafka-consumer.js:382]: Here you should use `DEFAULT_CONSUME_TIME_OUT` instead of `DEFAULT_CONSUME_LOOP_TIMEOUT_DELAY`

File: src/kafka-consumer.cc [hunk 11]
```diff
@@ -1044,16 +1044,16 @@ NAN_METHOD(KafkaConsumer::NodeResume) {
...
+  if (info.Length() < 3) {
     // Just throw an exception
     return Nan::ThrowError("Invalid number of parameters");
   }
 
-  if (!info[0]->IsNumber()) {
-    return Nan::ThrowError("Need to specify a timeout");
+  if (!info[1]->IsNumber()) {
+    return Nan::ThrowError("Need to specify a sleep delay");
   }
```
Comment (iradul) [inline src/kafka-consumer.cc:1054]: We need both `info[0]` and `info[1]`:
```
  if (!info[0]->IsNumber()) {
    return Nan::ThrowError("Need to specify a timeout");
  }

  if (!info[1]->IsNumber()) {
    return Nan::ThrowError("Need to specify a sleep delay");
  }
```

**Comment by iradul** [comment]
Thanks!

**Comment by yunnysunny** [comment]
I'm sorry for taking some mistakes. I wrote the code during my overtime work when my head was not clear.

**Comment by iradul** [comment]
No worries, thanks again for PR!
